firedrake/slate/slac.py

```diff
@@ -0,0 +1,473 @@
+    def get_c_str_bop(expr, temps, prec=None):
+        """Generates code representation of the binary
+        SLATE operations."""
+
+        op = {operator.add: '+',
+              operator.sub: '-',
+              operator.mul: '*'}[expr.op]
+        prec = expr.prec
+        result = "%s %s %s" % (get_c_str(expr.children[0], temps,
+                                         prec),
+                               op,
+                               get_c_str(expr.children[1], temps,
+                                         prec))
+
+        return pars(result, expr.prec, prec)
+
+    @get_c_str.register(Inverse)
+    def get_c_str_inv(expr, temps, prec=None):
+        """Generates code representation of the
+        inverse operation."""
+
+        return "(%s).inverse()" % get_c_str(expr.children, temps)
+
+    @get_c_str.register(Transpose)
+    def get_c_str_t(expr, temps, prec=None):
+        """Generates code representation of the
+        transpose operation."""
+
+        return "(%s).transpose()" % get_c_str(expr.children, temps)
+
+    # Creating c statement for the resulting linear algebra expression
+    result_type = map_type(mat_type(shape))
+    result_sym = ast.Symbol("T%d" % len(temps))
+    result_data_sym = ast.Symbol("A%d" % len(temps))
+    result = ast.Decl(dtype, ast.Symbol(result_data_sym, shape))
+
+    result_statement = ast.FlatBlock("%s %s((%s *)%s);\n" %
+                                     (result_type, result_sym,
+                                      dtype, result_data_sym))
+
+    statements.append(result_statement)
+    c_string = ast.FlatBlock(get_c_str(slate_expr, temps))
+    statements.append(ast.Assign(result_sym, c_string))
+
+    # Generating function arguments for macro kernel function
+    arglist = [result, ast.Decl("%s **" % dtype, coordsym)]
+    for c in coeffs:
+        ctype = "%s **" % dtype
+        if isinstance(c, firedrake.Constant):
+            ctype = "%s *" % dtype
+        arglist.append(ast.Decl(ctype, coeffmap[c]))
+
+    if need_cell_facets:
+        arglist.append(ast.Decl("char *", cellfacetsym))
+
+    kernel = ast.FunDecl("void", "slac_compile_slate", arglist,
+                         ast.Block(statements),
+                         pred=["static", "inline"])
+
+    # Finally we transform the kernel into a set of C++ template functions.
+    # This produces the final kernel AST
+    klist = []
+    transformkernel = Transformer()
+    oriented = False
+    for v in kernel_exprs.values():
+        for (_, ks) in v:
+            for k in ks:
+                oriented = oriented or k.kinfo.oriented
+                # TODO: Think about this. Is this true for SLATE?
+                assert k.kinfo.subdomain_id == "otherwise"
+                kast = transformkernel.visit(k.kinfo.kernel._ast)
+                klist.append(kast)
+
+    klist.append(kernel)
+    kernelast = ast.Node(klist)
+    inc.append('%s/lib/python2.7/site-packages/petsc/include/eigen3/' % sys.prefix)
```

Comment:
This is wrong if `firedrake-install` didn't install PETSc.  See how we define `get_petsc_dir` in `setup.py`.